File: optlib/convergence.py
"""Convergence and error-rate studies across the pricing engines.

The point of this module is to measure, not assert. For each engine it walks a
grid of discretization sizes, records signed error against a reference price and
the wall time taken, then fits the observed convergence order as the slope of
log|error| against log(n) by least squares.

Two caveats are surfaced in the output rather than smoothed over:

* Binomial error **oscillates**. Whether the terminal node lattice straddles the
  strike changes with n, so |error| has a sawtooth on top of its trend and the
  fitted slope has a genuinely poor R^2 unless payoff smoothing is on. The
  reported `r_squared` is how you tell a real rate from a fitted artefact.
* For Monte Carlo, error at one seed is a single draw from a distribution. The
  fitted slope is therefore reported next to the theoretical -1/2, and the
  standard error band is what should be compared, not the individual points.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

import numpy as np

from . import blackscholes as bs
from . import lattice, montecarlo as mc

# Engines available to the convergence study, keyed by the UI's identifier.
LATTICE_ENGINES = {
    "crr": {"label": "Binomial CRR", "method": "crr"},
    "crr_smooth": {"label": "CRR + payoff smoothing", "method": "crr", "smoothing": True},
    "crr_richardson": {"label": "CRR + smoothing + Richardson", "method": "crr", "smoothing": True, "richardson": True},
    "jarrow_rudd": {"label": "Binomial Jarrow-Rudd", "method": "jarrow_rudd"},
    "tian": {"label": "Binomial Tian", "method": "tian"},
    "leisen_reimer": {"label": "Binomial Leisen-Reimer", "method": "leisen_reimer"},
    "trinomial": {"label": "Trinomial (Boyle)", "trinomial": True},
}

MC_ENGINES = {
    "plain": {"label": "MC plain", "antithetic": False, "control_variate": False},
    "antithetic": {"label": "MC antithetic", "antithetic": True, "control_variate": False},
    "control": {"label": "MC control variate", "antithetic": False, "control_variate": True},
    "both": {"label": "MC antithetic + control", "antithetic": True, "control_variate": True},
    "qmc": {"label": "QMC Halton + both", "antithetic": True, "control_variate": True, "sampler": "halton"},
}


def _fit_order(sizes: np.ndarray, errs: np.ndarray) -> dict:
    """Fit |error| ~ C * n^(-p) by least squares in log-log space."""
    ok = np.isfinite(errs) & (np.abs(errs) > 0) & np.isfinite(sizes) & (sizes > 0)
    if ok.sum() < 3:
        return {"order": float("nan"), "r_squared": float("nan"), "n_points": int(ok.sum())}
    x = np.log(sizes[ok])
    y = np.log(np.abs(errs[ok]))
    A = np.column_stack([x, np.ones(x.size)])
    coef, *_ = np.linalg.lstsq(A, y, rcond=None)
    pred = A @ coef
    ss_res = float(np.sum((y - pred) ** 2))
    ss_tot = float(np.sum((y - y.mean()) ** 2))
    return {
        "order": float(-coef[0]),
        "intercept": float(coef[1]),
        "r_squared": float(1.0 - ss_res / ss_tot) if ss_tot > 0 else float("nan"),
        "n_points": int(ok.sum()),
    }


def lattice_convergence(
    S: float,
    K: float,
    T: float,
    r: float,
    q: float,
    sigma: float,
    is_call: bool = True,
    american: bool = False,
    engines: tuple[str, ...] = ("crr", "crr_smooth", "leisen_reimer", "trinomial"),
    n_min: int = 4,
    n_max: int = 800,
    n_points: int = 26,
) -> dict:
    """Error and timing versus step count for the lattice engines."""
    sizes = np.unique(np.round(np.geomspace(max(n_min, 2), max(n_max, n_min + 1), n_points)).astype(int))

    if american:
        reference = lattice.american_reference(S, K, T, r, q, sigma, is_call)
        ref_label = "Leisen-Reimer n=6001 + Richardson"
    else:
        reference = float(np.asarray(bs.price(S, K, T, r, q, sigma, is_call)).reshape(()))
        ref_label = "Black-Scholes closed form"

    series = []
    for key in engines:
        spec = LATTICE_ENGINES.get(key)
        if spec is None:
            continue
        pts = []
        for n in sizes:
            t0 = time.perf_counter()
            if spec.get("trinomial"):
                res = lattice.trinomial(S, K, T, r, q, sigma, int(n), is_call, american)
            else:
                res = lattice.binomial(
                    S, K, T, r, q, sigma, int(n), is_call, american,
                    spec["method"], spec.get("smoothing", False), spec.get("richardson", False),
                )
            ms = (time.perf_counter() - t0) * 1e3
            pts.append({"n": int(res.steps or n), "price": res.price, "error": res.price - reference, "ms": ms})
        errs = np.array([p["error"] for p in pts])
        ns = np.array([p["n"] for p in pts], dtype=float)
        series.append({
            "key": key,
            "label": spec["label"],
            "points": pts,
            "fit": _fit_order(ns, errs),
            "final_error": float(errs[-1]),
            "total_ms": float(sum(p["ms"] for p in pts)),
        })

    return {
        "kind": "lattice",
        "reference": reference,
        "reference_label": ref_label,
        "exercise": "american" if american else "european",
        "sizes": sizes.tolist(),
        "series": series,
        "note": (
            "Unsmoothed binomial error oscillates with n because the terminal lattice "
            "straddles the strike differently at each step count; a low R^2 on the "
            "fitted order is that sawtooth, not a bad fit."
        ),
    }


def mc_convergence(
    S: float,
    K: float,
    T: float,
    r: float,
    q: float,
    sigma: float,
    is_call: bool = True,
    engines: tuple[str, ...] = ("plain", "antithetic", "both", "qmc"),
    n_min: int = 512,
    n_max: int = 262_144,
    n_points: int = 10,
    seed: int = 987,
) -> dict:
    """Error, standard error and timing versus path count for the MC estimators."""
    sizes = np.unique(np.round(np.geomspace(max(n_min, 64), max(n_max, n_min * 2), n_points)).astype(int))
    reference = float(np.asarray(bs.price(S, K, T, r, q, sigma, is_call)).reshape(()))

    series = []
    for key in engines:
        spec = MC_ENGINES.get(key)
        if spec is None:
            continue
        pts = []
        for i, n in enumerate(sizes):
            t0 = time.perf_counter()
            res = mc.european(
                S, K, T, r, q, sigma, int(n), is_call,
                antithetic=spec.get("antithetic", False),
                control_variate=spec.get("control_variate", False),
                sampler=spec.get("sampler", "pseudo"),
                seed=seed + 101 * i,
                want_greeks=False,
            )
            ms = (time.perf_counter() - t0) * 1e3
            pts.append({
                "n": int(res.paths),
                "price": res.price,
                "error": res.price - reference,
                "std_error": res.std_error,
                "ms": ms,
                "efficiency": res.variance_reduction,
            })
        errs = np.array([p["error"] for p in pts])
        ns = np.array([p["n"] for p in pts], dtype=float)
        series.append({
            "key": key,
            "label": spec["label"],
            "points": pts,
            "fit": _fit_order(ns, errs),
            "se_fit": _fit_order(ns, np.array([p["std_error"] for p in pts])),
            "final_error": float(errs[-1]),
            "total_ms": float(sum(p["ms"] for p in pts)),
        })

    return {
        "kind": "monte-carlo",
        "reference": reference,
        "reference_label": "Black-Scholes closed form",
        "theoretical_order": 0.5,
        "sizes": sizes.tolist(),
        "series": series,
        "note": (
            "Plain Monte Carlo error decays as n^-1/2 regardless of variance reduction; "
            "what antithetic sampling and the control variate change is the constant, "
            "not the rate. Randomized-Halton QMC changes the rate itself."
        ),
    }


def engine_shootout(
    S: float,
    K: float,
    T: float,
    r: float,
    q: float,
    sigma: float,
    is_call: bool = True,
    american: bool = False,
    target_bp: float = 1.0,
) -> dict:
    """Cost of reaching a target accuracy, per engine.

    This is the comparison that actually matters in practice: not "which is more
    accurate at n = 500" but "what does each engine cost to get within a basis
    point of the premium". Reported as the wall time at the smallest
    discretization that hits the target and stays there.
    """
    if american:
        reference = lattice.american_reference(S, K, T, r, q, sigma, is_call)
    else:
        reference = float(np.asarray(bs.price(S, K, T, r, q, sigma, is_call)).reshape(()))
    tol = abs(reference) * target_bp / 10_000.0
    tol = max(tol, 1e-8)

    rows = []
    for key, spec in LATTICE_ENGINES.items():
        hit = None
        for n in (8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096):
            t0 = time.perf_counter()
            if spec.get("trinomial"):
                res = lattice.trinomial(S, K, T, r, q, sigma, n, is_call, american)
            else:
                res = lattice.binomial(
                    S, K, T, r, q, sigma, n, is_call, american,
                    spec["method"], spec.get("smoothing", False), spec.get("richardson", False),
                )
            ms = (time.perf_counter() - t0) * 1e3
            if abs(res.price - reference) <= tol:
                hit = {"n": int(res.steps or n), "ms": ms, "error": res.price - reference}
                break
        rows.append({
            "key": key, "label": spec["label"], "family": "lattice",
            "reached": hit is not None,
            "n": hit["n"] if hit else None,
            "ms": hit["ms"] if hit else None,
            "error": hit["error"] if hit else None,
        })

    if not american:
        for key, spec in MC_ENGINES.items():
            hit = None
            for n in (1024, 4096, 16_384, 65_536, 262_144, 1_048_576):
                t0 = time.perf_counter()
                res = mc.european(
                    S, K, T, r, q, sigma, n, is_call,
                    antithetic=spec.get("antithetic", False),
                    control_variate=spec.get("control_variate", False),
                    sampler=spec.get("sampler", "pseudo"),
                    seed=4242, want_greeks=False,
                )
                ms = (time.perf_counter() - t0) * 1e3
                # Require the confidence interval to sit inside the tolerance, not
                # just the point estimate: a lucky draw is not convergence.
                if abs(res.price - reference) <= tol and res.std_error * 1.96 <= tol * 2.0:
                    hit = {"n": int(res.paths), "ms": ms, "error": res.price - reference}
                    break
            rows.append({
                "key": key, "label": spec["label"], "family": "monte-carlo",
                "reached": hit is not None,
                "n": hit["n"] if hit else None,
                "ms": hit["ms"] if hit else None,
                "error": hit["error"] if hit else None,
            })

    return {
        "reference": reference,
        "target_bp": target_bp,
        "tolerance": tol,
        "exercise": "american" if american else "european",
        "rows": rows,
        "note": "Monte Carlo rows additionally require the 95% interval to fit the tolerance.",
    }

File: web/js/views/convergence.js
/* Convergence: measured error rates and the cost of hitting a target accuracy. */

import { el, clear, buildTable, fmtSig, fmtSigned, fmtMs, fmtInt, fmtCompact,
         seriesColor, token } from '../util.js';
import { api } from '../api.js';
import { state, contract, signature } from '../state.js';
import { LineChart, BarChart, withTableToggle } from '../chart.js';

const LATTICE_CHOICES = [
  ['crr', 'CRR'],
  ['crr_smooth', 'CRR + smoothing'],
  ['crr_richardson', 'CRR + smooth + Richardson'],
  ['jarrow_rudd', 'Jarrow-Rudd'],
  ['tian', 'Tian'],
  ['leisen_reimer', 'Leisen-Reimer'],
  ['trinomial', 'Trinomial'],
];

// Short forms for the endpoint labels; the legend carries the full name.
const SHORT = {
  crr: 'CRR', crr_smooth: 'CRR+sm', crr_richardson: 'CRR+sm+Rich',
  jarrow_rudd: 'J-R', tian: 'Tian', leisen_reimer: 'L-R', trinomial: 'Trinomial',
  plain: 'plain', antithetic: 'anti', control: 'control', both: 'anti+ctrl', qmc: 'QMC',
};

const MC_CHOICES = [
  ['plain', 'Plain MC'],
  ['antithetic', 'Antithetic'],
  ['control', 'Control variate'],
  ['both', 'Antithetic + control'],
  ['qmc', 'Halton QMC'],
];

export class ConvergenceView {
  constructor(host, ctx) {
    this.host = host;
    this.ctx = ctx;
    this.dirty = true;
    // Capped at five so categorical hues are never cycled.
    this.latticeEngines = new Set(['crr', 'crr_smooth', 'leisen_reimer', 'trinomial']);
    this.mcEngines = new Set(['plain', 'antithetic', 'both', 'qmc']);
    this.nMax = 800;
    this.pathsMax = 262144;
    this.targetBp = 1;
    this._build();
  }

  _build() {
    clear(this.host);
    this.notice = el('div');

    this.runBtn = el('button', { class: 'btn btn-primary', type: 'button', onclick: () => this.refresh().catch((e) => this.ctx.reportError(e, 'convergence')) }, 'Run study');

    this.latticeChips = el('div', { class: 'chip-row' });
    this.mcChips = el('div', { class: 'chip-row' });
    this._renderChips();

    const controls = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Study configuration'),
          el('p', {}, 'Each engine is walked over a geometric grid of discretization sizes; the '
            + 'convergence order is the fitted slope of log|error| against log(n).')),
        el('div', { class: 'card-tools' }, this.runBtn)),
      el('div', { class: 'panel-title' }, 'Lattice engines (max 5)'),
      this.latticeChips,
      el('div', { class: 'panel-title', style: { marginTop: '12px' } }, 'Monte Carlo estimators (max 5)'),
      this.mcChips,
      el('div', { class: 'row', style: { marginTop: '12px' } },
        el('label', { class: 'inline-label' }, 'max steps'),
        el('input', { type: 'number', value: String(this.nMax), min: '16', max: '4000', step: '100',
          style: { width: '92px' }, oninput: (e) => { this.nMax = Number(e.target.value) || 800; } }),
        el('label', { class: 'inline-label' }, 'max paths'),
        el('input', { type: 'number', value: String(this.pathsMax), min: '1024', max: '2000000', step: '10000',
          style: { width: '112px' }, oninput: (e) => { this.pathsMax = Number(e.target.value) || 262144; } }),
        el('label', { class: 'inline-label' }, 'target accuracy (bp)'),
        el('input', { type: 'number', value: String(this.targetBp), min: '0.01', max: '1000', step: '0.5',
          style: { width: '80px' }, oninput: (e) => { this.targetBp = Number(e.target.value) || 1; } })));

    this.latticeHost = el('div');
    this.latticeLegend = el('div');
    this.latticeTools = el('div', { class: 'card-tools' });
    this.latticeFit = el('div');
    this.latticeCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Lattice convergence'),
          this.latticeNote = el('p', {}, '')),
        this.latticeTools),
      this.latticeHost, this.latticeLegend, this.latticeFit);

    this.mcHost = el('div');
    this.mcLegend = el('div');
    this.mcTools = el('div', { class: 'card-tools' });
    this.mcFit = el('div');
    this.mcCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Monte Carlo convergence'),
          this.mcNote = el('p', {}, '')),
        this.mcTools),
      this.mcHost, this.mcLegend, this.mcFit);

    this.shootHost = el('div');
    this.shootTable = el('div');
    this.shootCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Cost to reach the target accuracy'),
          this.shootNote = el('p', {}, 'Wall time at the smallest discretization that lands inside the '
            + 'tolerance. Monte Carlo rows must also fit their 95% interval inside it, so a lucky draw '
            + 'does not count as convergence.')),
        el('div', { class: 'card-tools' })),
      this.shootHost, this.shootTable);

    this.host.append(this.notice, controls, this.latticeCard, this.mcCard, this.shootCard);

    this.latticeChart = new LineChart(this.latticeHost, {
      height: 320, legendHost: this.latticeLegend, xLabel: 'steps (log)', yLabel: '|error| (log)',
    });
    this.mcChart = new LineChart(this.mcHost, {
      height: 320, legendHost: this.mcLegend, xLabel: 'paths (log)', yLabel: '|error| (log)',
    });
    this.shootChart = new BarChart(this.shootHost, { xLabel: 'milliseconds (log)' });

    this.latticeTableToggle = withTableToggle(this.latticeChart, this.latticeTools);
    this.latticeCard.append(this.latticeTableToggle.holder);
    this.mcTableToggle = withTableToggle(this.mcChart, this.mcTools);
    this.mcCard.append(this.mcTableToggle.holder);
  }

  _renderChips() {
    const build = (host, choices, set) => {
      clear(host);
      for (const [key, label] of choices) {
        const on = set.has(key);
        host.append(el('button', {
          class: 'chip', type: 'button',
          style: on ? { borderColor: 'var(--accent)', background: 'var(--accent-soft)', color: 'var(--text-primary)' } : {},
          'aria-pressed': String(on),
          onclick: () => {
            if (set.has(key)) set.delete(key);
            else if (set.size >= 5) {
              this.ctx.toast('Five series is the cap — categorical hues are never cycled past the palette.', 'info');
              return;
            } else set.add(key);
            this._renderChips();
          },
        }, label));
      }
    };
    build(this.latticeChips, LATTICE_CHOICES, this.latticeEngines);
    build(this.mcChips, MC_CHOICES, this.mcEngines);
  }

  async refresh() {
    this.dirty = false;
    const sig = signature();
    [this.latticeChart, this.mcChart].forEach((c) => c.setStale(true));
    try {
      const res = await api.convergence({
        ...contract(),
        american: state.american,
        lattice_engines: [...this.latticeEngines],
        mc_engines: [...this.mcEngines],
        n_max: this.nMax,
        paths_max: this.pathsMax,
        target_bp: this.targetBp,
      });
      if (sig !== signature()) return;
      this.data = res;
      this._render();
    } finally {
      [this.latticeChart, this.mcChart].forEach((c) => c.setStale(false));
    }
  }

  _render() {
    const d = this.data;
    clear(this.notice);
    if (state.american) {
      this.notice.append(el('div', { class: 'notice info' }, el('div', {},
        el('strong', {}, 'American exercise. '),
        'There is no closed form, so error is measured against a high-accuracy lattice '
        + '(Leisen-Reimer at 6001 steps with Richardson extrapolation). The Monte Carlo panel is '
        + 'European-only — Longstaff-Schwartz is biased low by construction, which makes a '
        + 'convergence-rate fit against a fixed reference misleading.')));
    }

    this._renderLattice(d.lattice);
    this._renderMC(d.monte_carlo);
    this._renderShootout(d.shootout);
  }

  _renderLattice(L) {
    this.latticeCard.style.display = L ? '' : 'none';
    if (!L) return;
    const xs = L.series[0] ? L.series[0].points.map((p) => p.n) : [];
    this.latticeChart.setData({
      x: xs,
      xLog: true,
      yLog: true,
      series: L.series.map((s, i) => ({
        key: s.key,
        label: `${s.label} — order ${s.fit.order.toFixed(2)}`,
        shortLabel: SHORT[s.key] || s.label,
        values: s.points.map((p) => Math.abs(p.error)),
        color: seriesColor(i),
        markers: xs.length <= 30,
        fmt: (v) => fmtSig(v, 3),
      })),
      xName: 'steps',
      xTitle: (v) => `${fmtInt(v)} steps`,
      xTableFmt: (v) => fmtInt(v),
      valueFmt: (v) => fmtSig(v, 3),
    });
    this.latticeTableToggle.refresh();
    this.latticeNote.textContent = `Absolute pricing error against ${L.reference_label} `
      + `= ${fmtSig(L.reference, 8)}. Both axes are logarithmic, so a straight line is a power law and `
      + `its slope is the convergence order.`;

    const cols = [
      { label: 'Engine', mono: false, get: (s, i) => el('span', {},
        el('span', { class: 'swatch', style: { background: seriesColor(L.series.indexOf(s)) } }), s.label) },
      { label: 'Fitted order', get: (s) => s.fit.order.toFixed(3), cls: (s) => (s.fit.order > 1.5 ? 'pos' : '') },
      { label: 'R² of fit', get: (s) => s.fit.r_squared.toFixed(3),
        cls: (s) => (s.fit.r_squared < 0.8 ? 'neg' : '') },
      { label: `Error at n=${fmtInt(this.nMax)}`, get: (s) => fmtSigned(s.final_error, 3) },
      { label: 'Total time', get: (s) => fmtMs(s.total_ms) },
    ];
    clear(this.latticeFit).append(
      buildTable(cols, L.series),
      el('p', { class: 'axis-note' }, L.note));
  }

  _renderMC(M) {
    this.mcCard.style.display = M ? '' : 'none';
    if (!M) return;
    const xs = M.series[0] ? M.series[0].points.map((p) => p.n) : [];
    this.mcChart.setData({
      x: xs,
      xLog: true,
      yLog: true,
      series: M.series.map((s, i) => ({
        key: s.key,
        label: `${s.label} — order ${s.fit.order.toFixed(2)}`,
        shortLabel: SHORT[s.key] || s.label,
        values: s.points.map((p) => Math.abs(p.error)),
        color: seriesColor(i),
        markers: true,
        fmt: (v) => fmtSig(v, 3),
      })),
      xName: 'paths',
      xTitle: (v) => `${fmtCompact(v)} paths`,
      xTableFmt: (v) => fmtInt(v),
      valueFmt: (v) => fmtSig(v, 3),
    });
    this.mcTableToggle.refresh();
    this.mcNote.textContent = `Absolute error against the closed form = ${fmtSig(M.reference, 8)}. `
      + `Theoretical order for pseudorandom Monte Carlo is 0.50 regardless of variance reduction.`;

    const cols = [
      { label: 'Estimator', mono: false, get: (s) => el('span', {},
        el('span', { class: 'swatch', style: { background: seriesColor(M.series.indexOf(s)) } }), s.label) },
      { label: 'Fitted order', get: (s) => s.fit.order.toFixed(3), cls: (s) => (s.fit.order > 0.8 ? 'pos' : '') },
      { label: 'R² of fit', get: (s) => s.fit.r_squared.toFixed(3) },
      { label: 'Std-err order', get: (s) => s.se_fit.order.toFixed(3) },
      { label: 'Final error', get: (s) => fmtSigned(s.final_error, 4) },
      { label: 'Efficiency', get: (s) => {
        const last = s.points[s.points.length - 1];
        return last && Number.isFinite(last.efficiency) ? `${fmtSig(last.efficiency, 3)}×` : '—';
      } },
      { label: 'Total time', get: (s) => fmtMs(s.total_ms) },
    ];
    clear(this.mcFit).append(
      buildTable(cols, M.series),
      el('p', { class: 'axis-note' }, M.note,
        ' The standard-error column is the cleaner measurement: a single error point is one draw from a '
        + 'distribution, whereas the standard error is an estimate of that distribution\'s width.'));
  }

  _renderShootout(S) {
    if (!S) { this.shootCard.style.display = 'none'; return; }
    this.shootCard.style.display = '';
    const familyColor = (f) => (f === 'lattice' ? seriesColor(1) : seriesColor(2));
    this.shootChart.setData({
      xLog: true,
      valueFmt: (v) => fmtMs(v),
      bars: S.rows.map((r) => ({
        label: r.label,
        value: r.reached ? Math.max(r.ms, 0.01) : NaN,
        color: familyColor(r.family),
        note: r.reached ? `${fmtInt(r.n)} ${r.family === 'lattice' ? 'steps' : 'paths'}` : 'never reached the target',
      })),
    });

    const cols = [
      { label: 'Engine', mono: false, get: (r) => el('span', {},
        el('span', { class: 'swatch', style: { background: familyColor(r.family) } }), r.label) },
      { label: 'Reached', mono: false, get: (r) => el('span', { class: `tag ${r.reached ? 'good' : 'bad'}` }, r.reached ? '✓' : '✕') },
      { label: 'Resolution', get: (r) => (r.reached ? `${fmtInt(r.n)} ${r.family === 'lattice' ? 'steps' : 'paths'}` : '—') },
      { label: 'Time', get: (r) => (r.reached ? fmtMs(r.ms) : '—') },
      { label: 'Error there', get: (r) => (r.reached ? fmtSigned(r.error, 3) : '—') },
    ];
    clear(this.shootTable).append(
      buildTable(cols, S.rows),
      el('p', { class: 'axis-note' },
        `Target ${S.target_bp} bp of premium = ${fmtSig(S.tolerance, 3)} in price terms, against a `
        + `reference of ${fmtSig(S.reference, 8)}. `, S.note));
  }

  redraw() {
    [this.latticeChart, this.mcChart, this.shootChart].forEach((c) => c && c.draw());
    if (this.data) this._render();
  }
}
